src/linker/Linker/TypeNameResolver.cs

```diff
@@ -31,13 +31,37 @@ public TypeReference ResolveTypeName (string typeNameString)
 				return ResolveTypeName (null, assemblyQualifiedTypeName);
 			}
 
-			foreach (var assemblyDefinition in _context.GetReferencedAssemblies ()) {
-				var foundType = ResolveTypeName (assemblyDefinition, parsedTypeName);
-				if (foundType != null)
-					return foundType;
-			}
+			// If parsedTypeName doesn't have an assembly name in it but it does have a namespace,
+			// search for the type in the calling object's assembly, mscorlib and SPCL, in that order.
+			AssemblyDefinition assemblyOfCaller = origin switch {
+				TypeReference tr => tr.Module.Assembly,
+				_ => origin.DeclaringType.Module.Assembly
+			};
+
+			if (assemblyOfCaller != null && TryResolveTypeName (assemblyOfCaller, parsedTypeName, out var typeRef))
+				return typeRef;
+
+			AssemblyDefinition mscorlibAssembly = _context.GetLoadedAssembly ("mscorlib");
```

Comment:
`Type.GetType` only searches in whatever is the core library of the given runtime. The docs are confusing because they refer to both mscorlib and system.private.corelib, but the former is for .NET Framework, the latter is for Core.